### PR TITLE
8309603: Deallocate hashtables in ClassLoaderData::unload

### PR DESCRIPTION
**Pull request by chadrako**
[JDK-8309603](https://bugs.openjdk.org/browse/JDK-8309603)

Memory is released in `ClassLoaderdata::~ClassLoaderData` for the dictionary and module hashtables but they could be deleted earlier in `ClassLoaderdata::unload`. Since we're unloading the class loader is no longer being used for lookup.

Additional testing:
- [x] Linux x86_64 fastdebug `tier2`
- [x] Linux x86_64 release `tier2`
- [x] Linux x86_64 fastdebug `gtest:all`
- [x] Linux x86_64 release `gtest:all`
- [x] Linux x86_64 fastdebug `test/hotspot/jtreg/runtime`
- [x] Linux x86_64 release `test/hotspot/jtreg/runtime`

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309603](https://bugs.openjdk.org/browse/JDK-8309603): Deallocate hashtables in ClassLoaderData::unload (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14820/head:pull/14820` \
`$ git checkout pull/14820`

Update a local copy of the PR: \
`$ git checkout pull/14820` \
`$ git pull https://git.openjdk.org/jdk.git pull/14820/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14820`

View PR using the GUI difftool: \
`$ git pr show -t 14820`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14820.diff">https://git.openjdk.org/jdk/pull/14820.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14820#issuecomment-1631455423)

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 1]
```diff
@@ -734,20 +748,6 @@ ClassLoaderData::~ClassLoaderData() {
     _packages = nullptr;
   }
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderData.cpp:749]: Can you do packages too?

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 2]
```diff
@@ -734,20 +748,6 @@ ClassLoaderData::~ClassLoaderData() {
...
-  }
-
-  // Release C heap allocated hashtable for the dictionary
-  if (_dictionary != nullptr) {
-    // Destroy the table itself
-    delete _dictionary;
-    _dictionary = nullptr;
-  }
-
   if (_unnamed_module != nullptr) {
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderData.cpp:751]: And maybe unnamed module?

**Review by coleenp — CHANGES_REQUESTED** [review]

**Comment by dholmes-ora** [comment]
When do we delete the ClassLoaderData? Does moving this to unload time help with a specific problem? It seems odd to split the deallocation operations between unload and the destructor - why does unload not just call the destructor?

**Comment by coleenp** [comment]
Unload cannot just call the destructor because it's not safe to release the metaspace chunk yet, until the next handshake operation.  Since ZGenerational may take a while to request a handshake, this might be able to release some memory beforehand.  @fisk asked for this.
Did releasing the package table not work?

**Comment by chadrako** [comment]
Started failing a handful of tier1/tier2 tests after moving packages and unnamed module to unload. Running a couple more tests to see what the issue is

**Comment by chadrako** [comment]
> Started failing a handful of tier1/tier2 tests after moving packages and unnamed module to unload. Running a couple more tests to see what the issue is

Failed tests passed when ran again. They do not appear to be related

**Review by coleenp — APPROVED** [review]
This looks good.  While it's nicer to remove these in the destructor, this releases memory more eagerly for class unloading without waiting for a safepoint.

**Comment by fisk** [comment]
The design for concurrent class unloading was to essentially do this:
1. Unlink unloading data
2. Handshake all threads potentially accessing unloading data
3. Delete unloading data
The reason for this order, is that it provides safe memory reclamation. We want to do this consistently, so we don't get races where the code works 99.999999% of the time, except when you get a nasty race that crashes the JVM.
So unfortunately, this patch goes in the wrong direction, I think. I don't think we should do this.

**Pull request closed without merging**